### PR TITLE
Improve simulator type hints

### PR DESCRIPTION
**Pull request by alexrudd2**
Solves another 3 `mypy` errors.

`config_types` below seemed like a good candidate for using a `TypedDict` (~equivalent to a `struct`)
﻿https://github.com/pymodbus-dev/pymodbus/blob/214c7ed4ef27c686ad5d161b78477cc5aa6cb1cc/pymodbus/datastore/simulator.py#L110-L117

However, I ran into issues:
(1) `mypy` doesn't support using variables as `TypedDict` keys: https://github.com/python/mypy/issues/7178.  This unfortunate deficiency means `TypedDict` is not suitable here until `mypy` improves.
(2) Using `type` as a key is messy, since it's a ["soft keyword"](https://docs.python.org/3/reference/lexical_analysis.html#soft-keywords).  Maybe it would be better to use `kind`?  This should have a benefit beyond `mypy`.

### CODE REVIEW

File: pymodbus/datastore/simulator.py [hunk 1]
```diff
@@ -762,7 +762,7 @@ def validate_type(self, func_code, real_address, count):
             check = (CellType.UINT16, CellType.STRING)
             reg_step = 1
         else:
-            check = (CellType.UINT32, CellType.FLOAT32, CellType.STRING)
+            check = (CellType.UINT32, CellType.FLOAT32, CellType.STRING)  # type: ignore[assignment]
```
Comment (janiversen) [inline pymodbus/datastore/simulator.py:765]: That sounds wrong...that you need to ignore the assignment. Line 762 is ok....check contains the legal CellTypes and can be any number (of the valid types).

**Review by janiversen — CHANGES_REQUESTED** [review]

File: pymodbus/datastore/simulator.py (hunk 1 above)
Comment (alexrudd2) [reply pymodbus/datastore/simulator.py:765]: `mypy` does not like the fact that one path has a `tuple` with 2 items, and another with 3.

I don't see any meaningful consequence to the warning, so I ignored it.

Comment (alexrudd2) [reply pymodbus/datastore/simulator.py:765]: Do you prefer the other alternative of annotating `check: tuple` before any of the paths define it?

Comment (alexrudd2) [reply pymodbus/datastore/simulator.py:765]: > sadly enough it is not possible to annotate without assigning

Why do you say that?  It's perfectly possible with no assignment or `Non`

Comment (janiversen) [reply pymodbus/datastore/simulator.py:765]: ok, lets see what you can come up with....I thought annotation was only possible with assigning.

File: pymodbus/datastore/simulator.py [hunk 2]
```diff
@@ -748,11 +748,12 @@ def action_uptime(cls, registers, inx, cell, **_kwargs):
     # Internal helper methods
     # --------------------------------------------
 
-    def validate_type(self, func_code, real_address, count):
+    def validate_type(self, func_code, real_address, count) -> bool:
         """Check if request is done against correct type.
 
         :meta private:
         """
+        check: tuple
```
Comment (janiversen) [inline pymodbus/datastore/simulator.py:756]: That is surely good news, that did not work 1-2 years ago.

Thanks.

**Review by janiversen — APPROVED** [review]

**Pull request merged**